### PR TITLE
如何在 Keras 中用 YOLOv3 进行对象检测

### LINKED ISSUE #5900: 如何在 Keras 中用 YOLOv3 进行对象检测
* 原文链接：[How to Perform Object Detection With YOLOv3 in Keras](https://machinelearningmastery.com/how-to-perform-object-detection-with-yolov3-in-keras/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md)
  * PR 地址：https://github.com/xitu/gold-miner/pull/5961
  * 文章分类：**AI**
  * **注意：文件位置在 `TODO1` 文件夹中，不是之前的 `TODO` 了。**
  ----
  * 翻译时间：7 天
  * 校对时间：1-2 天
  * 翻译积分：7-8 分
  * 校对积分：1.5-4 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by leviding**
译文翻译完成，resolve #5900

### CODE REVIEW

**Comment by fanyijihua** [comment]
@zhmhhu 好的呢 :beer:

**Comment by leviding** [comment]
lsvih 校对 6 分

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 1]
```diff
@@ -2,138 +2,138 @@
...
+从头开始实现这个模型确实很有挑战性，特别是对新手来说，因为需要开发很多自定义的模型元素，来进行训练和预测。举个例子，即使是直接使用预训练过的模型，也需要复杂的代码来提炼和阐释模型输出的预测边边框
 
-Instead of developing this code from scratch, we can use a third-party implementation. There are many third-party implementations designed for using YOLO with Keras, and none appear to be standardized and designed to be used as a library.
+我们不从头写代码，而直接使用第三方实现过的。有很多第三方实现了为 Keras 的 YOLO，但没一个标准化的，也没一个设计得可用作库的。
 
-The [YAD2K project](https://github.com/allanzelener/YAD2K) was a de facto standard for YOLOv2 and provided scripts to convert the pre-trained weights into Keras format, use the pre-trained model to make predictions, and provided the code required to distill interpret the predicted bounding boxes. Many other third-party developers have used this code as a starting point and updated it to support YOLOv3.
+[YAD2K 项目](https://github.com/allanzelener/YAD2K) 是事实意义上的 YOLOv2 标准，而且提供了从预训练权值到 Keras 格式的转换脚本，来使用预训练的模型进行预测。此外还提供了代码，来提炼和解释预测的变边边框很多其他第三方开发者使用此代码作为起点，将其更新到可以支持 YOLOv3。
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: [YAD2K 项目](https://github.com/allanzelener/YAD2K) 是事实意义上的 YOLOv2 标准，它提供了将预先训练的权重转换为 Keras 格式的脚本，使用预先训练的模型进行预测，并提供提取解释预测边界框所需的代码。许多其他第三方开发人员已将此代码用作起点，并对其进行了更新以支持 YOLOv3。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 2]
```diff
@@ -2,138 +2,138 @@
...
 
-Perhaps the most widely used project for using pre-trained the YOLO models is called “[keras-yolo3: Training and Detecting Objects with YOLO3](https://github.com/experiencor/keras-yolo3)” by [Huynh Ngoc Anh](https://www.linkedin.com/in/ngoca/) or experiencor. The code in the project has been made available under a permissive MIT open source license. Like YAD2K, it provides scripts to both load and use pre-trained YOLO models as well as transfer learning for developing YOLOv3 models on new datasets.
+可能用于预训练的 YOLO 模型最广泛使用的项目是 [keras-yolo3：使用YOLO3训练和检测物体](https://github.com/experiencor/keras-yolo3) 由  [Huynh Ngoc Anh ](https://www.linkedin.com/in/ngoca/)或 experiencor 开发。该项目中的代码已在 MIT 开源许可下提供。 与 YAD2K 一样，该项目提供了脚本，可用于加载和使用预训练的YOLO模型，也可用于开发在新数据集上的 YOLOv3 模型。
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 使用预训练的 YOLO 模型最广泛使用的项目可能就是是 "[keras-yolo3：使用YOLO3训练和检测物体]"了，该项目由[Huynh Ngoc Anh ](https://www.linkedin.com/in/ngoca/)开发，也可称他为 experiencor。该项目中的代码已在 MIT 开源许可下提供。 与 YAD2K 一样，该项目提供了可用于加载和使用预训练的 YOLO 模型的脚本，也可在新数据集上开发基于 YOLOv3 的迁移学习模型。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 3]
```diff
@@ -2,138 +2,138 @@
...
 
-He also has a [keras-yolo2](https://github.com/experiencor/keras-yolo2) project that provides similar code for YOLOv2 as well as detailed tutorials on how to use the code in the repository. The [keras-yolo3](https://github.com/experiencor/keras-yolo3) project appears to be an updated version of that project.
+experiencor 还有一个 [keras-yolo2](https://github.com/experiencor/keras-yolo2) 项目，里面的代码和 YOLOv2 很像，也有详细教程叫你如何使用这个仓库的代码。[keras-yolo3](https://github.com/experiencor/keras-yolo3) 似乎是这个项目的更新版。
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 叫你如何使用
=>教你如何使用

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 4]
```diff
@@ -2,138 +2,138 @@
...
 
-Interestingly, experiencor has used the model as the basis for some experiments and trained versions of the YOLOv3 on standard object detection problems such as a kangaroo dataset, racoon dataset, red blood cell detection, and others. He has listed model performance, provided the model weights for download and provided YouTube videos of model behavior. For example:
+有意思的是，experiencor 以这个模型为基础做了些实验，在诸如袋鼠数据集、racoon 数据集、红细胞检测等等标准对象检测问题上，训练了 YOOLOv3 的多种版本。他列出了模型表现，还给出可以下载的下载模型权重，甚至有模型行为的 YouTube 视频。比如：
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 他列出了模型表现，还给出可以下载的下载模型权重，甚至有模型行为的 YouTube 视频。
=>他列出了模型表现结果，还给出了模型权重以供下载，甚至还发布了展示模型表现结果的 YouTube 视频。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 5]
```diff
@@ -2,138 +2,138 @@
...
+## 教程概览
 
-This tutorial is divided into three parts; they are:
+本教程分为三个部分，分别是：
 
-1.  YOLO for Object Detection
-2.  Experiencor YOLO3 Project
-3.  Object Detection With YOLOv3
+1. 用于对象检测的 YOLO
+2. experiencor 的 YOLO3 项目
```
[the remark this answers is not in the document]
Comment (lsvih) [reply TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 这儿的 experiencor 是个人

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 6]
```diff
@@ -2,138 +2,138 @@
...
+该方法涉及单个深度卷积神经网络（最初是 GoogLeNet 的一个版本，后来更新了，称为基于 VGG 的 DarkNet），将输入分成单元网格，每个格直接预测边框和对象分类。结果就是，经过后处理步骤，大量候选边边框并成为最终预测。
 
-There are three main variations of the approach, at the time of writing; they are YOLOv1, YOLOv2, and YOLOv3. The first version proposed the general architecture, whereas the second version refined the design and made use of predefined anchor boxes to improve bounding box proposal, and version three further refined the model architecture and training process.
+在写本文时有三种主要变体：YOLOv1、YOLOv2、YOLOv3。第一个版本提出通用架构，第二个版本改进设计，并利用预定义锚框改进了边边框议，第三个版本进一步完善模型架构和训练过程。
 
-Although the accuracy of the models is close but not as good as Region-Based Convolutional Neural Networks (R-CNNs), they are popular for object detection because of their detection speed, often demonstrated in real-time on video or with camera feed input.
+虽然模型的准确性接近基于区域的卷积神经网络（R-CNN），但不如后者那么好，但由于它们的检测速度快，因此在对象检测中很受欢迎，通常可以在视频或摄像机的输入上实时显示。
```
[the remark this answers is not in the document]
Comment (lsvih) [reply TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 虽然模型的准确性接近基于区域的卷积神经网络（R-CNN），但不及后者
->
虽然模型的准确性略逊于基于区域的卷积神经网络（R-CNN）

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhmhhu) [reply TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 是的，翻译到后面发现 experiencor 指的是 YOLO3 的作者 Huynh Ngoc Anh。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 7]
```diff
@@ -470,27 +470,27 @@ for i in range(len(yhat)):
 	boxes += decode_netout(yhat[i][0], anchors[i], class_threshold, input_h, input_w)
 ```
 
-Next, the bounding boxes can be stretched back into the shape of the original image. This is helpful as it means that later we can plot the original image and draw the bounding boxes, hopefully detecting real objects.
+接下来要将边框拉伸至原来图像的形状。稍后将打印原始图像、画出边框、并有希望检测到真实对象，因此这一步会很有帮助。
 
-The experiencor script provides the _correct\_yolo\_boxes()_ function to perform this translation of bounding box coordinates, taking the list of bounding boxes, the original shape of our loaded photograph, and the shape of the input to the network as arguments. The coordinates of the bounding boxes are updated directly.
+由 experiencor 提供的脚本中有 **correct\_yolo\_boxes()** 函数，可以转换边框坐标、获取边框列表、一开始加载图片的形状、和网络中作为参数输入的形状。边框的坐标直接更新：
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 由 experiencor 提供的脚本中有 **correct\_yolo\_boxes()** 函数，可以转换边框坐标，把边界框列表、一开始加载的图片的原始形状以及网络中输入的形状作为参数。边界框的坐标直接更新。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 8]
```diff
@@ -470,27 +470,27 @@ for i in range(len(yhat)):
...
-The experiencor script provides this via the _do_nms()_ function that takes the list of bounding boxes and a threshold parameter. Rather than purging the overlapping boxes, their predicted probability for their overlapping class is cleared. This allows the boxes to remain and be used if they also detect another object type.
+该脚本通过 **do_nms()**  实现这一点，该函数的参数是边框列表和阈值。该函数整理的不是重叠的边框，而是重叠类的预测概率。这样如果检测到另外的对象类型，边框仍还可用。
 
 ```
 # suppress non-maximal boxes
 do_nms(boxes,  0.5)
 ```
 
-This will leave us with the same number of boxes, but only very few of interest. We can retrieve just those boxes that strongly predict the presence of an object: that is are more than 60% confident. This can be achieved by enumerating over all boxes and checking the class prediction values. We can then look up the corresponding class label for the box and add it to the list. Each box must be considered for each class label, just in case the same box strongly predicts more than one object.
+这样留下的边框数量就一样了，但很无趣。于是就可以找到预测对象程度很强的边框，也就是有60%的置信度。通过遍历所有边框，核对预测值类型，可实现这一步步骤。然后找边框对应类型的标签，加入到列表中。每个边框需要跟每个类标签一一核对，这样可以避免同一边框强烈预测的对象不止一个的情况。
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 这样留下的边框数量就一样了，但只有少数有用。 我们只能检索那些强烈预测对象存在的边框：超过60％的置信率。 这可以通过遍历所有框并检查类预测值来实现。 然后，我们可以查找该框的相应类标签并将其添加到列表中。 每个边框需要跟每个类标签一一核对，以防同一个框强烈预测多个对象。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 9]
```diff
@@ -509,9 +509,9 @@ def get_boxes(boxes, labels, thresh):
 	return v_boxes, v_labels, v_scores
 ```
 
-We can call this function with our list of boxes.
+用边框列表当做参数调用该函数。
 
-We also need a list of strings containing the class labels known to the model in the correct order used during training, specifically those class labels from the MSCOCO dataset. Thankfully, this is provided in the experiencor script.
+需要一个字符串字表用于存放模型已知类标签，顺序要和训练模型时候的顺序保持一致，特别是来自 MSCOCO 数据集的类标签。庆幸的时候，这在脚本中也是有的。
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 我们还需要一个字符串列表，其中包含模型中已知的类标签，顺序要和训练模型时候的顺序保持一致，特别是 MSCOCO 数据集中的类标签。 值得庆幸的是，这些在 experiencor 的脚本中也提供。

File: TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md [hunk 10]
```diff
@@ -529,15 +529,15 @@ labels = ["person", "bicycle", "car", "motorbike", "aeroplane", "bus", "train",
...
+现在有了预测对象较强的少数边框，可以对它们做个总结。
 
 ```
 # summarize what we found
 for i in range(len(v_boxes)):
     print(v_labels[i], v_scores[i])
 ```
 
-We can also plot our original photograph and draw the bounding box around each detected object. This can be achieved by retrieving the coordinates from each bounding box and creating a Rectangle object.
+也可以打印原始图片，画出每个检测出对象的边框。通过从每个边框找到坐标，再创建Rectangle对象实现。
```
Comment (zhmhhu) [inline TODO1/how-to-perform-object-detection-with-yolov3-in-keras.md]: 我们还可以绘制原始照片并在每个检测到的物体周围绘制边界框。 这可以通过从每个边界框检索坐标并创建 Rectangle 对象来实现。
